File: src/commands/tasks.rs
use crate::client::AttioClient;
use crate::error::Result;
use crate::output::{print_many, print_one, OutputFormat};
use crate::types::{DataResponse, ListResponse, Task};
use serde::Serialize;

/// List tasks.
pub async fn list(
    client: &AttioClient,
    limit: Option<u32>,
    offset: Option<u32>,
    json: bool,
) -> Result<()> {
    #[derive(Serialize)]
    struct Query {
        #[serde(skip_serializing_if = "Option::is_none")]
        limit: Option<u32>,
        #[serde(skip_serializing_if = "Option::is_none")]
        offset: Option<u32>,
    }

    let query = Query { limit, offset };
    let response: ListResponse<Task> = client.post("/tasks/query", &query).await?;
    print_many(&response.data, OutputFormat::from_json_flag(json))?;
    Ok(())
}

/// Get a specific task.
pub async fn get(client: &AttioClient, task_id: &str, json: bool) -> Result<()> {
    let response: DataResponse<Task> = client.get(&format!("/tasks/{}", task_id)).await?;
    print_one(&response.data, OutputFormat::from_json_flag(json))?;
    Ok(())
}

/// Create a new task.
pub async fn create(
    client: &AttioClient,
    content: &str,
    deadline: Option<&str>,
    assignees: Option<&str>,
    linked_records: Option<&str>,
    json: bool,
) -> Result<()> {
    #[derive(Serialize)]
    struct CreateRequest {
        data: CreateData,
    }

    #[derive(Serialize)]
    struct CreateData {
        content: String,
        format: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        deadline_at: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        assignees: Option<serde_json::Value>,
        #[serde(skip_serializing_if = "Option::is_none")]
        linked_records: Option<serde_json::Value>,
    }

    let assignees: Option<serde_json::Value> = assignees.map(serde_json::from_str).transpose()?;
    let linked_records: Option<serde_json::Value> =
        linked_records.map(serde_json::from_str).transpose()?;

    let request = CreateRequest {
        data: CreateData {
            content: content.to_string(),
            format: "plaintext".to_string(),
            deadline_at: deadline.map(|s| s.to_string()),
            assignees,
            linked_records,
        },
    };

    let response: DataResponse<Task> = client.post("/tasks", &request).await?;
    print_one(&response.data, OutputFormat::from_json_flag(json))?;
    Ok(())
}

/// Update a task.
pub async fn update(
    client: &AttioClient,
    task_id: &str,
    content: Option<&str>,
    completed: Option<bool>,
    deadline: Option<&str>,
    json: bool,
) -> Result<()> {
    #[derive(Serialize)]
    struct UpdateRequest {
        data: UpdateData,
    }

    #[derive(Serialize)]
    struct UpdateData {
        #[serde(skip_serializing_if = "Option::is_none")]
        content: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        format: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        is_completed: Option<bool>,
        #[serde(skip_serializing_if = "Option::is_none")]
        deadline_at: Option<String>,
    }

    let request = UpdateRequest {
        data: UpdateData {
            content: content.map(|s| s.to_string()),
            format: content.map(|_| "plaintext".to_string()),
            is_completed: completed,
            deadline_at: deadline.map(|s| s.to_string()),
        },
    };

    let response: DataResponse<Task> = client
        .patch(&format!("/tasks/{}", task_id), &request)
        .await?;
    print_one(&response.data, OutputFormat::from_json_flag(json))?;
    Ok(())
}

File: src/attio_cli/commands/tasks.py
"""Task commands."""

import click

from attio_cli.cli_core import TASKS_CREATE_EXAMPLES, AttioCommand
from attio_cli.columns import TASK_COLUMNS
from attio_cli.commands.common import get_client
from attio_cli.output import get_json_input, output_many, output_one


@click.group()
def tasks() -> None:
    """Manage tasks."""


@tasks.command("list")
@click.option("--limit", type=int, help="Maximum number of tasks")
@click.option("--offset", type=int, help="Number of tasks to skip")
@click.option("--json", "as_json", is_flag=True, help="Output as JSONL")
def tasks_list(limit: int, offset: int, as_json: bool) -> None:
    """List tasks."""
    query = {}
    if limit:
        query["limit"] = limit
    if offset:
        query["offset"] = offset

    with get_client() as client:
        response = client.post("/tasks/query", query)
        output_many(response["data"], TASK_COLUMNS, as_json)


@tasks.command("retrieve")
@click.argument("task_id")
@click.option("--json", "as_json", is_flag=True, help="Output as JSON")
def tasks_retrieve(task_id: str, as_json: bool) -> None:
    """Retrieve a task by ID."""
    with get_client() as client:
        response = client.get(f"/tasks/{task_id}")
        output_one(response["data"], TASK_COLUMNS, as_json)


@tasks.command("create", cls=AttioCommand, examples=TASKS_CREATE_EXAMPLES)
@click.argument("content")
@click.option("--deadline", help="Deadline (ISO 8601 format)")
@click.option("--assignees", help="Assignees as JSON array")
@click.option("--assignees-file", help="Read assignees JSON from file or '-' for stdin")
@click.option("--linked-records", help="Linked records as JSON array")
@click.option(
    "--linked-records-file",
    help="Read linked records JSON from file or '-' for stdin",
)
@click.option("--json", "as_json", is_flag=True, help="Output as JSON")
def tasks_create(
    content: str,
    deadline: str,
    assignees: str,
    assignees_file: str,
    linked_records: str,
    linked_records_file: str,
    as_json: bool,
) -> None:
    """Create a new task."""
    data = {"content": content, "format": "plaintext"}
    if deadline:
        data["deadline_at"] = deadline
    assignee_value = get_json_input(
        data=assignees,
        data_file=assignees_file,
        required=False,
        context="assignees JSON",
        allow_stdin=False,
    )
    if assignee_value is not None:
        data["assignees"] = assignee_value
    linked_record_value = get_json_input(
        data=linked_records,
        data_file=linked_records_file,
        required=False,
        context="linked records JSON",
        allow_stdin=False,
    )
    if linked_record_value is not None:
        data["linked_records"] = linked_record_value

    with get_client() as client:
        response = client.post("/tasks", {"data": data})
        output_one(response["data"], TASK_COLUMNS, as_json)


@tasks.command("update")
@click.argument("task_id")
@click.option("--content", help="New task content")
@click.option("--completed", type=bool, help="Mark as completed")
@click.option("--deadline", help="New deadline (ISO 8601 format)")
@click.option("--json", "as_json", is_flag=True, help="Output as JSON")
def tasks_update(task_id: str, content: str, completed: bool, deadline: str, as_json: bool) -> None:
    """Update a task."""
    data = {}
    if content is not None:
        data["content"] = content
        data["format"] = "plaintext"
    if completed is not None:
        data["is_completed"] = completed
    if deadline is not None:
        data["deadline_at"] = deadline

    with get_client() as client:
        response = client.patch(f"/tasks/{task_id}", {"data": data})
        output_one(response["data"], TASK_COLUMNS, as_json)
